port/unix/omrsysinfo.c

```diff
@@ -3071,3 +3158,356 @@ omrsysinfo_os_kernel_info(struct OMRPortLibrary *portLibrary, struct OMROSKernel
+		 * 	5:cpuacct,cpu,cpuset:/daemons
+		 * The colon-separated fields are, from left to right:
+		 * 	1. hierarchy ID number
+		 * 	2. set of subsystems bound to the hierarchy
+		 * 	3. control group in the hierarchy to which the process belongs
+		 */
+		rc = fscanf(cgroupFile, "%d:%[^:]:%s", &hierId, subsystems, cgroup);
+		if (EOF == rc) {
+			rc = 0;
+			break;
+		} else if (3 != rc) {
+			rc = portLibrary->error_set_last_error_with_message(portLibrary, OMRPORT_ERROR_SYSINFO_PROCESS_CGROUP_FILE_READ_FAILED, "unexpcted format of process cgroup file");
+			goto _end;
+		}
+		cursor = subsystems;
+		do {
+			const char *cgName = NULL;
+
+			separator = strchr(cursor, ',');
+			if (NULL != separator) {
+				*separator = '\0';
+			}
+
+			cgName = getCgroupNameForSubsystem(portLibrary, cgEntryList, cursor);
+
+			if (NULL == cgName) {
+				rc = addCgroupEntry(portLibrary, &cgEntryList, hierId, cursor, cgroup);
+				if (0 != rc) {
+					goto _end;
+				}
+			} else {
+				char errBuf[512];
+
+				portLibrary->str_printf(portLibrary, errBuf, sizeof(errBuf), "multiple entries for subsystem %s found in process cgroup file", cursor);
+				rc = portLibrary->error_set_last_error_with_message(portLibrary, OMRPORT_ERROR_SYSINFO_PROCESS_CGROUP_FILE_DUPLICATE_SUBSYSTEM_ENTRIES, errBuf);
+				goto _end;
+
+			}
+			if (NULL != separator) {
+				cursor = separator + 1;
+			}
+		} while (NULL != separator);
+	}
+	rc = 0;
+
+_end:
+	if (NULL != cgroupFile) {
+		fclose(cgroupFile);
+	}
+	if (0 != rc) {
+		freeCgroupEntries(portLibrary, cgEntryList);
+		cgEntryList = NULL;
+	} else {
+		*cgroupEntryList = cgEntryList;
+	}
+	
+	return rc;
+}
+
+/**
+ * Read a file under a subsystem in cgroup hierarchy based on the format specified
+ *
+ * @param[in] portLibrary pointer to OMRPortLibrary
+ * @param[in] subsystem cgroup subsystem
+ * @param[in] fileName name of the file under cgroup subsystem
+ * @param[in] numItemsToRead number of items to be read from the file
+ * @param[in] format format to be used for reading the file
+ *
+ * @return 0 on successfully reading 'numItemsToRead' items from the file, negative error code on any error
+ */
+static int32_t
+readCgroupSubsystemFile(struct OMRPortLibrary *portLibrary, CgroupSubsystems subsystem, const char *fileName, int32_t numItemsToRead, const char *format, ...)
+{
+	char fileToRead[PATH_MAX];
+	char *cgroup = NULL;
+	FILE *file = NULL;
+	int32_t rc = 0;
+	va_list args;
+
+	if (!portLibrary->sysinfo_is_cgroup_limits_enabled(portLibrary)) {
+		rc = portLibrary->error_set_last_error_with_message(portLibrary, OMRPORT_ERROR_SYSINFO_CGROUP_LIMITS_DISABLED, "cgroup limits is disabled");
+		goto _end;
+	}
+	
+	cgroup = getCgroupNameForSubsystem(portLibrary, PPG_cgroupEntryList, subsystemNames[subsystem]);
+	if (NULL == cgroup) {
+		char errBuf[512];
+
+		portLibrary->str_printf(portLibrary, errBuf, sizeof(errBuf), "cgroup name for subsystem %s is not available", subsystemNames[MEMORY]);
```

Comment:
I'll whine about checking for truncation here as well but its less of a concern as the subsystem names are more constrained.